_published/0012-authentication.md

```diff
@@ -0,0 +1,167 @@
+Or for a more fun version checkout [this game by auth0](https://iverab-my.sharepoint.com/:i:/g/personal/fredrik_dyrvold_iver_se/Ef41bR71qeNOk-QWhCar0V4BZN8WqvM7iEf5g7ml9_s7Ew?e=ZoZRN6)
+
+### Identity provider
+
+We should connect to azure ad on our local deployments.
+
+We should not hav our own identity provider setup with wharf for now.
+
+We should also be able to publish code in such a way that the IDP is
+interchangeable. If it is used by the public they can set up their own IDP
+and connect to that.
+
+### Authorisation server
+
+There are many [alternatives](https://openid.net/developers/certified/) on
+alternative authorisation servers.
+
+We had a quick info share from spark the spark project. They are using open
+identity server 3 or 4. This however appears to be relatively tightly coupled
+with the .NET framework.
+
+One that seems to align well with this
+project is `Ory - Hydra` [github](https://github.com/ory/hydra)
+and [docs](https://www.ory.sh/hydra/docs/). The descriptions of them from their
+github at the present time is:
+
+```
+ORY Hydra is a hardened, OpenID Certified OAuth 2.0 Server and OpenID Connect
+Provider optimized for low-latency, high throughput, and low resource
+consumption. ORY Hydra is not an identity provider (user sign up, user
+login, password reset flow), but connects to your existing identity provider
+through a login and consent app. Implementing the login and consent app in a
+different language is easy, and exemplary consent apps (Node) and SDKs for all
+common languages are provided.
+```
+They appear to be funded through opencollective and patreon.
+
+Another alternative to consider could be [Keycloak](https://www.keycloak.org/).
+This appears to contain
+much more finished work such as a user admin panel for
+setting user and group rights. It is styled using themes. Con: Its a redhat
+project which means association with IBM which has had a certain downside in
+relation to opensource projects in the past see
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any identity provider (IDP) can be used.
+Locking ourselves to a single IDP when a standard such as OICD exists seems
+unnecessary. Keeping the choice of IDP open will leave more room for
+choice in the future.
+
+## Alternative solutions
+
+The starting assumption for this RFC is usage of the [OpenID Connect spec](https://openid.net/specs/openid-connect-core-1_0.html).
+There is a [list of certified providers](https://openid.net/developers/certified/).
+
+## Future possibilities
+
+After this we should consider
+- User management interfaces
+- User group management interfaces
+- Authorization
+- Permissions
+
+## Unresolved questions
```

Comment:
I know we'll get the question further on about LDAP or just using a in-database account store. How does this solution handle this?

Some of the best services, from the user and admin perspective, that we use are for example Grafana and GitLab that allows us to set up LDAP authentication as well as creating custom local users.

How does this proposed ORY Hydra solution handle this? For example mixing multiple login providers, such as username+password (for the initial admin account), an OIDC login (such as Azure AD or via GitHub Apps), and a LDAP login, all side by side (if the operations team so wanted it). Maybe even login from multiple LDAP or OIDC connections?

I sense that this is difficult to change later on if we end up only supporting OIDC and nothing else.

Reply:
I think its possible to read microsoft style AD FS ldap attributes through OIDC scopes. There is an [article here](https://cloud.google.com/anthos/clusters/docs/on-prem/1.7/how-to/oidc-adfs#mapping_ldap_attributes_to_claim_names) that outlines a plan for how its done. Testing this for our use case which seems quite different requires setting up infrastructure that has costs (in both time an money) associated to it. I do not intend to test this in practice as of now.
If we use OIDC to connect an AD FS IdP to our hydra server then we are likely using LDAP i assume the same goes for azureAD. If we connect hydra to kratos (both are ory products) then we would not be using LDAP. Kratos uses postgres by default but has some other options.
If you happen to mean SAML instead of LDAP then the answer is no we would not support SAML as far as this RFC goes. The implementation is different. Use of OIDC is a starting assumption for this RFC. This is stated in the summary.

Multiple different OIDC endpoints are possible.

> I sense that this is difficult to change later on if we end up only supporting OIDC and nothing else.

Yes. I belive adding support for other authentication styles would be difficult counter productive if we achive compatability with OIDC. Use of OIDC is a starting assumption for this RFC. This is stated in the summary.

Reply:
Good answer!

So for the sake of this RFC, we could ignore LDAP. But supporting username+password locally, which as I understand it is possible with ORY Hydra, is more essential even if this is the initial proposal

Reply:
This particular scenario has changed quite a bit since the rewrite. In the new model i propse we include a flipswitch to disable token validation in the backend for developer environments. This could allow navigation as normal entierly without logging in.